Implement umfMemoryProviderAllocation[Split/Merge]

igchor opened the pull request:
That is needed to support jemalloc (avoid leaking memory).

@kswiecicki could you verify if this solves all the memory leaks? On my machine, it seems to work fine after those changes.

pbalcer commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
     return ret;
 }
 
+static umf_result_t trackingAllocationSplit(void *provider, void *ptr,
```
Given that we use a tree for tracking, it might be simpler to avoid preemptively splitting/merging pages, and instead handle that during inserts/removals.

vinser52 replied:
it might be error-prone concerning IPC handle support. I prefer to do it explicitly when a pool manager calls merge/split functions (like implemented right now).
In general, it is an interesting case, the IPC handle is created per coarse-grain allocation (returned by the memory provider) and if two coarse-grain allocations are merged what should we do with corresponding IPC handles if them were created before? I will think more about that. I know that L0 just does not support split/merge functionality but tracking memory provider should work for all cases.

igchor replied:
Yeah, I was thinking about this as well. Doing this lazily in removal would definitely be simpler (and probably faster) but as Sergey said, might be problematic for IPC. I think we can revisit this once we have IPC support merged.

pbalcer commented on include/umf/memory_provider_ops.h:
```diff
@@ -139,6 +139,11 @@ typedef struct umf_memory_provider_ops_t {
+
+    umf_result_t (*allocation_split)(void *provider, void *ptr,
+                                     size_t totalSize, size_t leftSize);
+    umf_result_t (*allocation_merge)(void *provider, void *leftPtr,
```
you are making and assumption that "left" means "numerically lower address", right? I'd call it `lowPtr`, `highPtr`.

pbalcer reviewed (commented):
as for concurrency, I think if we don't add a lock we are creating a footgun. I can imagine a scenario where for example, split is called concurrently twice on the same block.

So unless, we add some mechanism for exclusive ownerships of ranges of memory, I think we should add a lock.

vinser52 commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+
+    tracker_value_t *splitValue =
```
if you allocate `splitValue` just before `critnib_insert` then you do not need to `free(splitValue)` in every error cases below.

vinser52 commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+        (umf_tracking_memory_provider_t *)provider;
+
+    tracker_value_t *mergedValue =
+        (tracker_value_t *)malloc(sizeof(tracker_value_t));
```
if you allocate `mergedValue` just before `critnib_insert` then you do not need to `free(mergedValue)` in every error case below.

vinser52 replied:
Make sense, but in case of merge do we really need to allocate a new element? You can just update the left record with the new size and remove the right record. Did I miss something?

igchor replied:
See my other answer about updating the values. Without this alloc we would have to modify the value in-place which means we would have to use atomic instructions.

vinser52 commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+        (tracker_value_t *)critnib_get((critnib *)p->hTracker, (uintptr_t)ptr);
+    if (!value) {
+        free(splitValue);
+        fprintf(stderr, "tracking split: no such value\n");
```
I would add `TODO` comment that logger should be called instead of just `fprintf`

vinser52 commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    }
+    if (value->size != totalSize) {
+        free(splitValue);
+        fprintf(stderr, "tracking split: %zu != %zu\n", value->size, totalSize);
```
I would add `TODO` comment that logger should be called instead of just `fprintf`

vinser52 commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+                                                        totalSize, leftSize);
+    if (ret != UMF_RESULT_SUCCESS) {
+        free(splitValue);
+        fprintf(stderr,
```
I would add `TODO` comment that logger should be called instead of just `fprintf`

vinser52 commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+        (critnib *)p->hTracker, (uintptr_t)leftPtr);
+    if (!leftValue) {
+        free(mergedValue);
+        fprintf(stderr, "tracking merge: no left value\n");
```
I would add `TODO` comment that logger should be called instead of just `fprintf`

vinser52 commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+        (critnib *)p->hTracker, (uintptr_t)rightPtr);
+    if (!rightValue) {
+        free(mergedValue);
+        fprintf(stderr, "tracking merge: no right value\n");
```
I would add `TODO` comment that logger should be called instead of just `fprintf`

vinser52 commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    }
+    if (leftValue->pool != rightValue->pool) {
+        free(mergedValue);
+        fprintf(stderr, "tracking merge: pool mismatch\n");
```
I would add `TODO` comment that logger should be called instead of just `fprintf`

vinser52 commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    }
+    if (leftValue->size + rightValue->size != totalSize) {
+        // TODO: should we allow leftValue->size + rightValue->size > totalSize?
+        fprintf(stderr, "tracking merge: leftValue->size + rightValue->size != "
```
I would add `TODO` comment that logger should be called instead of just `fprintf`

igchor replied:
Hm, do we really want to add "TODO"s to every fprintf call? I think we have a few without annotations already. When we will finally have a logging framework we'll do `grep fprintf` anyway I guess.

vinser52 commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+    if (leftValue->size + rightValue->size != totalSize) {
+        // TODO: should we allow leftValue->size + rightValue->size > totalSize?
```
How it is possible?

igchor replied:
Well, someone (implementer of a pool) can call split with wrong arguments (size that does not match existing regions). It's a programmer error but the check is pretty cheap and I think quite useful.

ldorau commented:
I confirm that the memory leak described in https://github.com/oneapi-src/unified-memory-framework/issues/147 does not occur with this change, because `arena_extent_split()` succeeds now.

kswiecicki commented:
> @kswiecicki could you verify if this solves all the memory leaks? On my machine, it seems to work fine after those changes.

It does, ASan no longer reports any leaks in jemalloc.

igchor commented:
> as for concurrency, I think if we don't add a lock we are creating a footgun. I can imagine a scenario where for example, split is called concurrently twice on the same block.
> 
> So unless we add some mechanism for exclusive ownerships of ranges of memory, I think we should add a lock.

Hm, should only add a mutex for split/merge calls or should we also protect trackingFree? I think freeing a memory region that is being split/merged can just be UB, right?

ldorau commented on include/umf/memory_provider_ops.h:
```diff
@@ -139,6 +139,11 @@ typedef struct umf_memory_provider_ops_t {
+
+    umf_result_t (*allocation_split)(void *provider, void *ptr,
+                                     size_t totalSize, size_t leftSize);
+    umf_result_t (*allocation_merge)(void *provider, void *leftPtr,
```
That's right :-)

ldorau commented:
This PR is required by #153

igchor commented:
> > as for concurrency, I think if we don't add a lock we are creating a footgun. I can imagine a scenario where for example, split is called concurrently twice on the same block.
> > So unless we add some mechanism for exclusive ownerships of ranges of memory, I think we should add a lock.
> 
> Hm, should only add a mutex for split/merge calls or should we also protect trackingFree? I think freeing a memory region that is being split/merged can just be UB, right?

I've added mutex for split/merge only for now.

ldorau commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    }
+
+    void *highPtr = (void *)(((uintptr_t)ptr) + firstSize);
+    size_t rightSize = totalSize - firstSize;
```
> size_t rightSize = totalSize - firstSize;

(`lowSize` and `highSize`) OR (`firstSize` and `secondSize`) OR (`leftSize` and `rightSize`) - do not mix them, please

bratpiorka commented on src/pool/pool_jemalloc.c:
```diff
@@ -210,15 +210,14 @@ static bool arena_extent_split(extent_hooks_t *extent_hooks, void *addr,
 
-    // TODO: add this function to the provider API to support Windows and USM
-    return true; // true means failure (unsupported)
+    assert(size_a + size_b == size);
```
please assert that size, size_a and size_b > 0 and addr != NULL

bratpiorka commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+
+    tracker_value_t *splitValue =
```
add a check for splitValue != NULL

ldorau commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    void *highPtr = (void *)(((uintptr_t)ptr) + firstSize);
+    size_t rightSize = totalSize - firstSize;
+
+    // We'll have duplicate entry for the range [highPtr, rightValue->size] but this is fine,
```
1) `have duplicate` -> `have to duplicate` ?
2) `highPtr, rightValue` ? - maybe `highPtr` and `highValue` ?

bratpiorka commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+                                            size_t totalSize,
+                                            size_t firstSize) {
+    umf_result_t ret = UMF_RESULT_ERROR_UNKNOWN;
+    umf_tracking_memory_provider_t *p =
```
could we stay with the full name? During the review I had to scroll up to see what is "p". If you want to keep the code short it would be better to add an alias for "tracker = provider->hTracker"

ldorau commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+
+    // We'll have duplicate entry for the range [highPtr, rightValue->size] but this is fine,
+    // the value is the same anyway and we forbid splitting/removing that range concurrently
+    ret = umfMemoryTrackerAdd(p->hTracker, p->pool, highPtr, rightSize);
```
`highPtr, rightSize` - mixed

ldorau commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    ret = umfMemoryTrackerAdd(p->hTracker, p->pool, highPtr, rightSize);
+    if (ret != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "tracking split: umfMemoryTrackerAdd failed\n");
+        // TODO: what now? should we rollback the split? This is can only happen due to ENOMEM
```
`This is can only happen` -> `This can only happen`

bratpiorka commented on include/umf/memory_provider.h:
```diff
@@ -156,6 +156,34 @@ umfMemoryProviderGetName(umf_memory_provider_handle_t hProvider);
+///        can be managed (freed) separately.
+/// @param hProvider handle to the memory provider
+/// @param ptr pointer to the beginning of the allocation
+/// @param totalSize total size of the allocation to be split
```
I see that in split/merge methods we don't want to have redundant arguments, so why we need "totalSize"? Shouldn't the totalSize be just the size of the alloc pointed by ptr? What if totalSize would be less/more than allocation?

vinser52 replied:
Here are my thoughts about this question:

1. We should not just copy jemalloc extent API because for me it looks jemalloc-specific. And it looks like we are not doing that.
2. Regarding `totalSize`. For me, it looks similar to `umfMemoryProviderFree(..., void *ptr, size_t size)` API that also requires passing the `size` of allocation pointed by `ptr`. But not all memory provider implementations need/use it: e.g. mmap implementation uses `size`, but L0 ignores it.
3. For me, bullet 2 means that UMF requires that the pool manager remember the size of each particular coarse-grain allocation but memory providers are not required to remember the size of each coarse-grain allocation. Since we already require to pass the `size` to `umfMemoryProviderFree` function what is the difference here? 
4. Regarding what if `totalSize != actual total size`, the same question is applicable for `umfMemoryProviderFree`. On our side we can do (at least debug) verification using the tracking memory provider - there is such information in the tracker.
5. Another question regarding input parameter validation, should the `firstSize` be a multiple of the page size?

bratpiorka replied:
ok I agree with bullet 3

igchor replied:
Regarding 5: I think it is implied because I'm saying that the pointer needs to point to the beginning of the allocation and all allocations need to be page aligned.

bratpiorka commented on include/umf/memory_provider.h:
```diff
@@ -156,6 +156,34 @@ umfMemoryProviderGetName(umf_memory_provider_handle_t hProvider);
+/// @param hProvider handle to the memory provider
+/// @param lowPtr pointer to the first allocation
+/// @param highPtr pointer to the second allocation (should be > lowPtr)
+/// @param totalSize size of a new merged allocation. Should be equal
```
why we need the totalSize argument?

ldorau commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+        (umf_tracking_memory_provider_t *)provider;
+
+    tracker_value_t *mergedValue =
+        (tracker_value_t *)malloc(sizeof(tracker_value_t));
```
handle malloc failure

ldorau commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    mergedValue->pool = p->pool;
+    mergedValue->size = totalSize;
+
+    if (!mergedValue) {
```
move that before the assignment

bratpiorka commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+        fprintf(stderr, "tracking split: umfMemoryTrackerAdd failed\n");
+        // TODO: what now? should we rollback the split? This is can only happen due to ENOMEM
+        // so it's unlikely but probably the best solution would be to try to preallocate everything
+        // (value and critnib nodes) before calling umfMemoryProviderAllocationSplit.
```
note: I added an issue for that

ldorau commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    }
+
+    tracker_value_t *leftValue =
+        (tracker_value_t *)critnib_get(p->hTracker->map, (uintptr_t)lowPtr);
```
`lowPtr` and `leftValue` - mixed names (low and high OR left and right)

ldorau commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+        goto err;
+    }
+    tracker_value_t *rightValue =
+        (tracker_value_t *)critnib_get(p->hTracker->map, (uintptr_t)highPtr);
```
`highPtr` and `rightValue` - mixed

ldorau commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+        goto err;
+    }
+
+    // We'll have duplicate entry for the range [highPtr, rightValue->size] but this is fine,
```
`have duplicate` -> `have to duplicate`

ldorau replied:
`highPtr, rightValue` - mixed

bratpiorka commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+                                            void *highPtr, size_t totalSize) {
+    umf_result_t ret = UMF_RESULT_ERROR_UNKNOWN;
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
```
it would be better to do "tracker = provider->hTracker"

ldorau commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    }
+
+    // We'll have duplicate entry for the range [highPtr, rightValue->size] but this is fine,
+    // the value is the same anyway and we forbid splitting/removing that range concurrently
```
splitting -> merging ?

igchor replied:
Just removing now, since split/merge is called under the lock.

bratpiorka commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+
+    tracker_value_t *mergedValue =
```
please check for "mergedValue != NULL"

igchor replied:
Done.

bratpiorka commented on src/provider/provider_tracking.h:
```diff
@@ -14,14 +14,41 @@
+}
+
+static inline void umfMemoryTrackerDestroy(umf_memory_tracker_handle_t handle) {
+    critnib_delete(handle->map);
```
add an assert + return if handle == NULL

igchor replied:
Can't return since this is done in the lib destructor anyway and we cannot handle a failure there.

bratpiorka replied:
Ok but if handle == NULL then we crash when we try to deref it. Please just add an early return for this case.

igchor replied:
Right, good point. Done.

igchor commented on src/pool/pool_jemalloc.c:
```diff
@@ -210,15 +210,14 @@ static bool arena_extent_split(extent_hooks_t *extent_hooks, void *addr,
+    assert(size_a + size_b == size);
+
+    jemalloc_memory_pool_t *pool = get_pool_by_arena_index(arena_ind);
+    return umfMemoryProviderAllocationSplit(pool->provider, addr, size,
```
Done.

igchor commented on src/pool/pool_jemalloc.c:
```diff
@@ -231,15 +230,12 @@ static bool arena_extent_merge(extent_hooks_t *extent_hooks, void *addr_a,
-    // TODO: add this function to the provider API to support Windows and USM
-    return true; // true means failure (unsupported)
+    jemalloc_memory_pool_t *pool = get_pool_by_arena_index(arena_ind);
+    return umfMemoryProviderAllocationMerge(pool->provider, addr_a, addr_b,
```
Done.

bratpiorka commented:
@vinser52 @pbalcer @ldorau all I need 2nd approve to merge this PR - could you check if anything is missing here?

vinser52 commented on src/provider/provider_tracking.h:
```diff
@@ -10,18 +10,61 @@
 
+struct umf_memory_tracker_t {
+    critnib *map;
+    os_mutex_t *splitMergeMutex;
```
Global split/merge lock might kill scalability. As I understand the lock is needed only when concurrent split/merge occurs for the same memory buffer.

pbalcer replied:
well, mmap is also fully synchronized, so I wouldn't worry that much. This shouldn't be a super common operation. I'd start with a simple mutex and then we can think about locking out ranges of memory if there's a perf problem.

pbalcer commented on src/provider/provider_tracking.c:
```diff
@@ -119,6 +121,164 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+                              secondSize);
+    if (ret != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "tracking split: umfMemoryTrackerAdd failed\n");
+        // TODO: what now? should we rollback the split? This can only happen due to ENOMEM
```
yes, preallocation makes sense - should be doable once we have the base allocator.
